### DISCUSSION TITLE
What is the recommended way to globally define macros?  (also, that take arguments)

### DISCUSSION THREAD

**Discussion by hrbigelow** [KaTeX/KaTeX#3166] [Q&A]
> Before requesting a feature, please search for [existing issues](https://github.com/KaTeX/KaTeX/issues).

I was reading through [https://github.com/KaTeX/KaTeX/issues/2070](https://github.com/KaTeX/KaTeX/issues/2070) and have successfully used the technique mentioned there to define global macros that take arguments:

```javascript
    let macros = {
      '\\B': '\\boldsymbol{#1}',
      '\\proj': '#1_{\\small \\parallel #2}'
    };
    document.addEventListener("DOMContentLoaded", function() {
        renderMathInElement(document.body, {
          macros: macros,
          
          // customised options
          // • auto-render specific keys, e.g.:
...
```

Since the post is a few years old, I wonderedi f there some HOWTO or update that I should also read.

Thanks very much!

**Comment by edemaine** [comment]
Thanks for checking in.  That is still the recommended method.  In the future, you might use the Discussions tab.

**Reply by hrbigelow** [reply]
Great, thanks.  Actually, would it also be possible to achieve this at the content level, where Katex first parses (but doesn't render anything for) `\\newcommand` statements like:

```javascript
    katex.renderToString(
        String.raw`\newcommand\proj[2]{#1_{\small \parallel #2}}`
    )
```

so that the global `options.macros` object gets updated?  I saw this other post:  [https://github.com/KaTeX/KaTeX/issues/3112](https://github.com/KaTeX/KaTeX/issues/3112) but it looks still open so not sure what the status is.

Thanks again

**Reply by edemaine** [reply]
You can use
```js
const macros = {};
katex.renderToString(
        String.raw`\newcommand\proj[2]{#1_{\small \parallel #2}}`,
        {macros, globalGroup: true});
// render other things with macros
```
Note that you need to pass in the same `macros` object to all KaTeX calls in order to share macros.

**Reply by hrbigelow** [reply]
Hi, this technique worked (Thank you!).  But I was wondering if there is any way to tell Katex not to render anything when it processes macro *definitions*.  At the moment, I see:

```javascript
const macros = {};
katex.renderToString(
        String.raw`\newcommand\B[1]{\boldsymbol{#1}}`,
        { macros, globalGroup: true });
// '<span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow></mrow><annotation encoding="application/x-tex">\\newcommand\\B[1]{\\boldsymbol{#1}}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"></span></span>' 
```

**Reply by edemaine** [reply]
`katex.renderToString` is just returning a string; it doesn't output anything.  Because you're running it in console, you're seeing the return value in console, but if you run this in a script or otherwise, there won't be any output.

**Reply by hrbigelow** [reply]
Oh, sorry, I shouldn't have used the word "output".  What I was really asking is, is there a way to tell katex.renderToString to return the empty string if it processes a macro definition, rather than returning the string as I've shown above?
